File: 2022/swift/13.swift
import Foundation

let filename = "./input/13.in"

enum Packet {
    case num(Int), arr([Packet])

    func compare(_ rhs: Packet) -> Int {
        let lhs = self
        switch (lhs, rhs) {
        case let (.num(left), .num(right)):
            return left < right ? 1 : left > right ? -1 : 0
        case (.num(_), .arr(_)):
            return Packet.arr([lhs]).compare(rhs)
        case (.arr(_), .num(_)):
            return lhs.compare(Packet.arr([rhs]))
        case let (.arr(left), .arr(right)):
            var pointer = 0
            while true {
                let isLeftOver = pointer == left.count
                let isRightOver = pointer == right.count

                if isLeftOver, isRightOver {
                    return 0
                }

                if isLeftOver {
                    return 1
                }

                if isRightOver {
                    return -1
                }

                let result = left[pointer].compare(right[pointer])

                if result != 0 {
                    return result
                }

                pointer += 1
            }
        }
    }

    static func == (lhs: Packet, rhs: Packet) -> Bool {
        switch (lhs, rhs) {
        case let (.num(left), .num(right)):
            return left == right
        case (.num(_), .arr(_)), (.arr(_), .num(_)):
            return false
        case let (.arr(left), .arr(right)):
            if left.count != right.count {
                return false
            }

            return left.enumerated().allSatisfy { $0.1 == right[$0.0] }
        }
    }
}

/*

 The decode function and JSON:Codable struct are all base off:

 https://adamrackis.dev/blog/swift-codable-any

 */

func decode(fromArray container: inout UnkeyedDecodingContainer) -> [Packet] {
    var result: [Packet] = []

    while !container.isAtEnd {
        if let value = try? container.decode(Int.self) {
            result.append(Packet.num(value))
        } else if var nestedArray = try? container.nestedUnkeyedContainer() {
            let decoded = decode(fromArray: &nestedArray)

            result.append(Packet.arr(decoded))
        }
    }

    return result
}

struct JSON: Codable {
    var value: Packet?

    init(from decoder: Decoder) throws {
        if var array = try? decoder.unkeyedContainer() {
            let decoded = decode(fromArray: &array)
            value = Packet.arr(decoded)
        } else if let value = try? decoder.singleValueContainer() {
            if let result = try? value.decode(Int.self) {
                self.value = Packet.num(result)
            }
        }
    }

    func encode(to _: Encoder) throws {}
}

func parse(_ json: String) -> Packet? {
    let decoder = JSONDecoder()
    let data = json.data(using: .utf8)!

    return try? decoder.decode(JSON.self, from: data).value
}

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let packets = contents.components(separatedBy: "\n\n").map {
            pair in
            pair.split(separator: "\n").map(String.init).map(parse).compactMap { $0 }
        }

        let partOne = packets.enumerated().filter {
            entry in
            let (_, pair) = entry
            let result = pair[0].compare(pair[1])
            return result == 1
        }.map { $0.0 + 1 }.reduce(0,+)

        print("Part one:", partOne)

        var flat = packets.flatMap { $0 }

        let twoMarker = parse("[[2]]")!
        let sixMarker = parse("[[6]]")!

        flat.append(twoMarker)
        flat.append(sixMarker)

        flat.sort {
            $0.compare($1) == 1
        }

        let twoPos = flat.firstIndex { $0 == twoMarker }
        let sixPos = flat.firstIndex { $0 == sixMarker }

        if let twoPos, let sixPos {
            print("Part two:", (twoPos + 1) * (sixPos + 1))
        }
    } catch {
        print(error)
    }
}

main()

File: 2022/src/bin/13.rs
use serde_json::Value;
use std::cmp::Ordering::{self, Equal, Greater, Less};

fn compare(left: &Value, right: &Value) -> Ordering {
    match (left, right) {
        (Value::Number(l), Value::Number(r)) => {
            let left_number = l.as_u64().unwrap();
            let right_number = r.as_u64().unwrap();

            return left_number.cmp(&right_number);
        }
        (Value::Array(l), Value::Array(r)) => {
            let mut pointer = 0;

            loop {
                let is_left_over = pointer == l.len();
                let is_right_over = pointer == r.len();

                if is_left_over && is_right_over {
                    return Equal;
                }

                if is_left_over {
                    return Less;
                }

                if is_right_over {
                    return Greater;
                }

                let result = compare(&l[pointer], &r[pointer]);

                if result != Equal {
                    return result;
                }

                pointer += 1;
            }
        }
        (Value::Number(_), Value::Array(_)) => {
            return compare(&Value::from(vec![left.clone()]), right);
        }
        (Value::Array(_), Value::Number(_)) => {
            return compare(left, &Value::from(vec![right.clone()]));
        }
        _ => panic!("Unexpected comparison"),
    }
}

#[aoc2022::main(13)]
fn main(input: &str) -> (usize, usize) {
    let pairs = input
        .split("\n\n")
        .map(|pair| {
            pair.split("\n")
                .map(|packet| {
                    let value: Value = serde_json::from_str(packet).unwrap();

                    return value;
                })
                .collect::<Vec<Value>>()
        })
        .collect::<Vec<Vec<Value>>>();

    let part_one = pairs
        .iter()
        .map(|pair| compare(&pair[0], &pair[1]))
        .enumerate()
        .filter(|(_, result)| *result == Less)
        .map(|(index, _)| index + 1)
        .sum::<usize>();

    let two_marker: Value = serde_json::from_str("[[2]]").unwrap();
    let six_marker: Value = serde_json::from_str("[[6]]").unwrap();

    let mut packets = pairs.iter().flatten().collect::<Vec<_>>();

    packets.push(&two_marker);
    packets.push(&six_marker);

    packets.sort_by(|a, b| compare(a, b));

    let two_pos = packets
        .iter()
        .position(|&packet| packet == &two_marker)
        .unwrap()
        + 1;

    let six_pos = packets
        .iter()
        .position(|&packet| packet == &six_marker)
        .unwrap()
        + 1;

    (part_one, two_pos * six_pos)
}
